Disable use of JSC signal based options when mach exception handler use is not allowed.

MenloDorian opened the pull request:
#### 1add1eb4c01a6511e0f681ec347d7d2aa2f1c140
<pre>
Disable use of JSC signal based options when mach exception handler use is not allowed.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278401">https://bugs.webkit.org/show_bug.cgi?id=278401</a>
<a href="https://rdar.apple.com/134053581">rdar://134053581</a>

Reviewed by Keith Miller and Yusuke Suzuki.

If the sandbox policy is known for mach exception handler use, we&apos;ll either allow or block it
in accordance with the sandbox policy.

if the sandbox policy is not known, we will use JIT disablement as a proxy indicator that use
of mach exception handler use is not allowed in the system.  So, in thos case, we&apos;ll disable
JSC signal based options when the JIT is disabled.

Note that all places where we disable the JIT (via setting useJIT to false) will be followed by
a call to Options::notifyOptionsChanged().  So, that is the natural place to disable signal
handler based options for this purpose if appropriate.

* Source/JavaScriptCore/API/tests/testapi.cpp:
(configureJSCForTesting):
* Source/JavaScriptCore/jsc.cpp:
(jscmain):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::disableAllSignalHandlerBasedOptions):
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/Options.h:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
(WebKit::XPCServiceInitializer):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::setJSCOptions):
* Source/WebKit/WebProcess/EntryPoint/Cocoa/XPCService/WebContentServiceEntryPoint.mm:
(WEBCONTENT_SERVICE_INITIALIZER):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(dumpRenderTree):

Canonical link: <a href="https://commits.webkit.org/282592@main">https://commits.webkit.org/282592@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e980518b4db2a61c53113358fa93fb217ea7225

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42933 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14185 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65697 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50621 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14465 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51194 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9815 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66646 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39846 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55057 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31879 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36529 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12432 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13058 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/56689 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58036 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12753 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69294 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62822 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7524 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12326 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58502 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7557 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55144 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58731 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6277 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/84583 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9621 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38754 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14905 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39833 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40945 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39576 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (approved):
r=me.

Constellation commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -676,6 +676,16 @@ static inline void disableAllJITOptions()
+    Options::useSharedArrayBuffer() = false;
+    Options::useWasmFastMemory() = false;
+    Options::useWasmFaultSignalHandler() = false;
+}
```
What happens when we are in WebContent process with LDM? We need to enable them.

Constellation reviewed (changes requested):
Yeah, I think we need to enable them on LDM. So this requires some more changes.

Constellation reviewed (approved):
r=me

Commit:
Disable use of JSC signal based options when mach exception handler use is not allowed.
https://bugs.webkit.org/show_bug.cgi?id=278401
rdar://134053581

Reviewed by Keith Miller and Yusuke Suzuki.

If the sandbox policy is known for mach exception handler use, we'll either allow or block it
in accordance with the sandbox policy.

if the sandbox policy is not known, we will use JIT disablement as a proxy indicator that use
of mach exception handler use is not allowed in the system.  So, in thos case, we'll disable
JSC signal based options when the JIT is disabled.

Note that all places where we disable the JIT (via setting useJIT to false) will be followed by
a call to Options::notifyOptionsChanged().  So, that is the natural place to disable signal
handler based options for this purpose if appropriate.

* Source/JavaScriptCore/API/tests/testapi.cpp:
(configureJSCForTesting):
* Source/JavaScriptCore/jsc.cpp:
(jscmain):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::disableAllSignalHandlerBasedOptions):
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/Options.h:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
(WebKit::XPCServiceInitializer):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::setJSCOptions):
* Source/WebKit/WebProcess/EntryPoint/Cocoa/XPCService/WebContentServiceEntryPoint.mm:
(WEBCONTENT_SERVICE_INITIALIZER):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(dumpRenderTree):

Canonical link: https://commits.webkit.org/282592@main
